[Manual Squash] Redesign/unify mainmenu settings interface

rubenwardy opened the pull request:
Fixes #12452, fixes #3266, fixes #12976, fixes #9016, fixes #12140, fixes #11069

The settings tab and all settings has been replaced with a single interface for changing settings

![image](https://user-images.githubusercontent.com/2122943/232167857-2baac6f6-c38c-431b-a464-20e8ee50cdc5.png)

## Technical intro

The dialog has a list of pages grouped into sections. When the dialog loads, it builds the pages by reading settingtypes.txt. 

When opening a page, the dialog builds components for each setting in the page. It's possible to define custom components in the setting menu, this is done to retain the current logic for the shader selection component.

A component is a table with a get_formspec function. Functions in components.lua build components for settings.

## To do

This PR is Ready for Review

- [x] Reimplement
- [x] Use scrollbar options to get proper scrollbars
- [x] Readd key changes
- [x] Show current page in leftpane
- [x] Restructure settingtypes.txt
- [x] Fix translation
- [x] Filter pages when searching
- [x] Validate min max
- [x] Readd reset
- [x] Restore element behaviour on "Most used" pane (what was the settings tab, eg: the shadow quality dropdown)
- [x] Readd show_technical_names
- [x] Fix issue with dropdown events (see todo in bool component)
- [x] Undersampling defaults to 0 in .cpp, but 1 in settingtypes.txt
- [x] Finish support for path select dialog
- [x] Improve styling 
- [x] The "Edit" button for the mapgen noises should be a bit wider (in German, the translation barely fits)
- [x] `1 ~= 1.0` reset button bug
- [x] Use image and tooltip area for info icons
- [x] Info icon for noise settings
- [x] Crash when editing a text field, eg: selectionbox_color
- [x] Icon still appears when there's no description, eg: font_bold
- [x] Larger box around show_technical_names
- [x] ~~Replace dummy tab with settings icon~~ (unnecessary work)
- [x] ~~Most Used > shader settings: Show technical names~~ (planning to remove Most Used in the future)


For future PRs (in rough order of priority):

- Fix dropdowns changing when scrolling ( https://github.com/minetest/irrlicht/pull/192 )
- Only show settings when relevant (ie: don't show shader settings when shaders are disabled), remove Most Used tab
- Add flag utils to Settings ref
- Sort mods based on title when show_technical_names=false
- Hide advanced settings by default (including mapgen settings)
- Noise dialog needs validation for Nan, inf, -inf.
- Noise dialog needs tooltips
- Support reset for noise parameters
- Improve keyboard use
- Mapgen flags translation
- Better use of font size (eg: of sub section headers)
- Support for drop down translation (hard coded)
- Add color type
- Use slider for some fields (like mouse sensitivity)
- Better way to show unchanged formspec ( https://github.com/minetest/minetest/issues/10432 )
- Limit font size / support larger fonts

## How to test

Types of setting: int, float, string, bool, v3f, flags, path, filepath, noise, enum

* Try editing one of each type. Make sure reset icon appears when relevant
* Try resetting one of each type
* Browse through, checking layout/structure/etc

<!-- Example code or instructions -->

nerzhul commented:
whereas i like the idea of the tab pane on the left and it's pretty nice, the double tab is clearly not convenient for the global UI. I think we should redesign the whole main pane model.
Maybe settings must have his full view like many games, and not be in the main menu tab, at least for this first version.

rubenwardy commented:
> whereas i like the idea of the tab pane on the left and it's pretty nice, the double tab is clearly not convenient for the global UI. I think we should redesign the whole main pane model. Maybe settings must have his full view like many games, and not be in the main menu tab, at least for this first version.

This is designed for the main menu redesign, which doesn't use tabs like we do now

What I could do is make this a modal dialog, and have the Settings tab open straight into it. Or I could just break convention and make the tab much bigger

sfan5 commented:
> What I could do is make this a modal dialog, and have the Settings tab open straight into it. Or I could just break convention and make the tab much bigger

Would be okay with me.

rubenwardy commented:
This is ready for review, although probably not ready to merge yet

If 5.6.0 is less than a month away, then I'd rather this waits for 5.7.0 as there's a lot of things left to refine

For example, here's a dependent settings feature: https://user-images.githubusercontent.com/2122943/176328135-4c65bc80-6039-4168-883f-0590d76bd952.mp4

rubenwardy commented:
Also, Zughy suggests that it be moved to a dialog accessed by a settings icon button:

![image](https://user-images.githubusercontent.com/2122943/176328202-7578a133-4b30-4a19-8185-cb2fe580fc83.png)

Zughy commented:
>  5.6.0 is less than a month away, then I'd rather this waits for 5.7.0 as there's a lot of things left to refine

Considering that it took you 2 days to do this, that I'm actively providing you feedback and graphical assets day by day, that the old settings are less user friendly than this anyway, and that we have a Minecraft situation out there we should bend in our favour, I think it's best to schedule this for 5.6; because right now a whole new main menu doesn't seem that far away for a 5.7 (seeing you working on it is fueling me up to finish the whole design). And we all know that the main menu is probably THE thing that keeps people away the most.

This doesn't have to be perfect, it just needs to work and not crash anything. Fanciness can come later (also because right now it has to stick with the design of the rest of menu)

> Rather than making the tab bigger

Yeah, if definitely breaks the flow having a bigger tab. This should make everyone in this conversation happy (ruben, sfan, me)

rubenwardy commented:
>  I think it's best to schedule this for 5.6

It's going to need someone to review it + subsequent PRs. I'm not that hopeful given how my other UX PRs wait around

Montandalar commented:
**UX Notes**
* No disrespect to Zughy's art abilities but a mix of pixel and smooth graphics is jarring. Unless we fully adopt the pixel aesthetic I would prefer higher resolution graphics for the reset and info icons.
* I'm not sure if prefixing "Content:" onto Mods and Games helps in any way
* I think the UX is still a little bit unclear about whether resetting a setting resets it to factory default or reverts the user's current changes, which might be exacerbated by some bugs.
* The interface is no longer as keyboard-friendly.
  * Keyboard-power-users like myself do fit in a weird space between direct minetest.conf editors and mouse-only operators, but I still edit my settings almost entirely in-game. I know keyboard controls is an afterthought for formspec, I don't think formspecs can even have a focused graphic to distinguish them while tabbing around, but 
  * The previous interface could basically be entirely navigated by keyboard, with expanding sections and so on, and by tabbing from the search box into the expanding sections. Hitting enter on the search bar would move between search results.
    * I think the search results should remove the tabbing at left and just put headings in line with the right hand pane. That would be quite keyboard-navigable.
  * Now you can't read the descriptions because they rely on mouseover to display as tooltips, and you have to hold the tab key for a good couple of seconds to tab from the sections at left to the current menu at right.
  * The info icon has a lesser tab index than the corresponding reset icon for any setting. Thus, instead of tabbing Item-Reset-Info as it appears visually, the tabbing goes Item-Info-Reset.
    * It would be good if the tooltip text could be activated on a formspec element by hitting space/enter while it's focused, or on the same timer as mouseover.

* I know the existing settings menu doesn't do this, but it would be much better to sort the Mod entries by their displayed name. This would also probably require a different sort conditional on whether show_technical_names is enabled (when that gets re-implemented.
  * Also related to show_technical_names: What about putting the technical name in the tooltip, at least in the interim?


* When scrolling with the mousewheel over the left side, if the mouse hovers over a dropdown, it will start scrolling options in that dropdown instead of continuing to scroll. This is exacerbated by the above mentioned bug with the reset buttons on dropdowns not appearing. This is probably not a regression, but this is the single most common interface in minetest where most people will have a scrollbar and dropdowns present.

**Interface Bugs**
**Dropdown menus (enum) don't allow returning to defaults correctly:**

    1. Clear any line that has `language =` in your minetest.conf
    2. Set language to anything other than "" in-game (I chose `de`) - the reset button will not appear next to the language dropdown as it should.
    3. Quit the game so the setting is saved. Re-open Minetest. The reset button appears now for the first time.
    4. Change the language setting to several others, including `""`. The reset button never disappears, even for the `""` setting. Change it back to `""` and close Minetest. You now have an entry `language = ` (yes, that has a trailing space) in your minetest.conf
    5. Restart Minetest. Now the reset button for language will never appear, which means it is acting the same as in step 2.
    6. Change the language setting to anything other than `""` again and restart again. Now press the reset button. Now switch it to various languages. No matter what you do with the language setting, the reset button will never appear again.

**Int and float**

  1. Clear any entry in minetest.conf with `undersampling = `
  2. Go to the graphics tab. Undersampling will have a value of 0 and a reset button, despite its default and range being 1 1 8 in settingtypes.txt
  3. Type -5 in the field and try to set it. It is clamped to 1. The reset button disappears because this is the default.
  4. Type -5 in the field again and set it. Close Minetest. The set button won't clamp the value, but the actual value that gets written to minetest.conf will be clamped properly.
  * Unlike integers, setting a float setting does clamp it as soon as you press set (I used GUI scaling)
  * Setting an integer setting (I used undersampling) to outside its range is not error-checked except in that when you leave the settings tab it gets clamped to the nearest in-range value.
* Typing anything other than a number in an int/float field and using the set button  will result in a big ugly error message.
* Entering `NaN` or `-Inf` in an int/float field will convert it to 0 automatically, and clamping to minimum will work from there as well.
* Entering `Inf` in an int/float field will convert it so it's clamped at the highest possible value for that setting.


**v3f** (I used mapgen fractal scale)
* Any permutation of capitalisation of `NaN` (`nan`, `NAN`) will get normalised to `nan` and accepted as valid.
* Any permutation of capitalisation of `Inf` (`inf`, `INF`) will get normalised to `inf` and accepted as valid.
* Good stuff? Negative zero is normalised to 0. Exponential notation works (e.g. 1e2 = 100).
* Excessively large positive values such as 1e4096 become inf; excessively large negatives become -inf.
* No feedback if input was invalid. It just ignores everything, doesn't show a reset button, and forgets the invalid value ever existed if you close and reopen Minetest.

**string**
* The crosshair colour defaults to black if the value is invalid as an RGB triplet, but the default in settingtypes.txt is white.
* Crosshair colour will dissappear if you attempt something like a formspec injection: `tooltip[0,0;8,8;pwned]`. Thankfully the reset button can fix this easily. I'm not sure if it's possible to do a proper injection since I couldn't get my tooltip to appear.

**bool**
There is some kind of issue with `bool` with default `false`, but only for mods? Sorry I couldn't reproduce this for any builtin settings such as mip mapping, despite the lines in the respect settingtypes files looking fine...

1. Install [foodblocks](https://github.com/Montandalar/foodblocks)
2. Remove any `foodblocks_*` settings from `minetest.conf`.
3. Start Minetest. Type `food` into the settings search and hit enter.
4. Despite being unchecked and the default false, the reset button appears.
5. Use the reset button. Toggle the checkbox as many times as you want.
7. Once some value is set in `minetest.conf`, the reset button appears as expected.
8. Close and reopen Minetest. Checkbox and reset button still work as expected.

**flags**
The same bug that effects `bool` seems to effect mapgen v6 flags, mapgen v7 flags, carpathian flags, flat flags and valleys flags: if you don't have the default values recorded in minetest.conf, it will offer a reset despite being at the default. Once the reset is pushed it adds the default in and only offers a reset when sensible.

**filepath**
I have this version installed through dpkg made by `cpack -G DEB`; it is not a `RUN_IN_PLACE` version.

Font paths:
1. Clear any `font_path_*` settings from `minetest.conf`
2. Open Minetest and go to the Advanced settings tab. The regular, bold, italic, bold&italic font paths all have all been set to `/usr/local/share/minetest/fonts/Arimo-*.ttf` and not the suggested resettable path of `fonts/Arimo-Regular.ttf` which is specified in settingtypes.txt. This would not be the correct relative path to the fonts from where my binary is located in `/usr/local/bin/minetest`; and surely it would be `../fonts/` instead for RUN_IN_PLACE builds of Minetest?

* /dev/null as a font path is very disappointing :P
* Setting the font path manually works fine

* Selecting a file path through the file browser is NOT working (I tested 5.5.1 and it works fine)
  1. Open Minetest and go to the Advanced settings tab
  2. Click the Browse button next to Regular font path
  3. Type/paste `/usr/share/fonts/truetype/liberation` (or some other path if you don't have those fonts or are on a different OS) and press enter.
  4. Click LiberationSerif-Regular.ttf
  5. Click OK
  6. Notice that the font path in the field has not actually changed

**noise**
Accepts Inf, -inf, NaN as valid values. Sets octaves and seed to 0 if invalid. I set most values in the heat noise (mg_biome_np_heat) to NaN and it results no soil at all, probably due to NaNs propagating.

Could at least defaults, eased, and absvalue, if not all parameters, get tooltips please?

Zughy commented:
> a mix of pixel and smooth graphics is jarring

You're right. I guess the general thought was "it's ugly anyway, so why not". It's also done in CDB window

> I'm not sure if prefixing "Content:" onto Mods and Games helps in any way

Agree, it's just redundant

> I think the UX is still a little bit unclear about whether resetting a setting resets it to factory default or reverts the user's current changes, which might be exacerbated by some bugs

Looks pretty straightforward to me

> Now you can't read the descriptions because they rely on mouseover to display as tooltips

Option names should be straightforward, I don't see the point in putting text everywhere. Tooltips exist for a reason

> I know the existing settings menu doesn't do this, but it would be much better to sort the Mod entries by their displayed name

Out of scope, and mods (as well as games) are not expected to remain there for long anyway

> When scrolling with the mousewheel over the left side, if the mouse hovers over a dropdown, it will start scrolling options in that dropdown instead of continuing to scroll

This should be fixed on a formspec level, it's not ruben's fault

In general, thank you *so much* for the in-depth review (I'll leave the rest to ruben I guess)!

Montandalar commented:
> Option names should be straightforward, I don't see the point in putting text everywhere. Tooltips exist for a reason

Well yes, but no that's not quite the point. In those bullet points I'm talking about how the *whole* interface should be keyboard-navigable, including the tooltips. Like the mod name sorting this is probably out of scope and should get filed under 'keyboard control improvements for formspec', which so far would be at least: (1) visual focus indication on buttons and so on (2) tooltips on keyboard focus.

>>I think the UX is still a little bit unclear ... which might be exacerbated by some bugs
>>
>Looks pretty straightforward to me

As long as we get the bugs ironed out there should be no confusion.

>In general, thank you so much for the in-depth review (I'll leave the rest to ruben I guess)!

It's easier to tear apart than to build, but anyway I wasn't going to let something like this which looks pretty important go without a proper testing :)

Extex101 commented:
Much needed and looks great 👍

rubenwardy commented:
Note: I've put "TODO" as reminders for me to add to the PR's to do list

> I think the UX is still a little bit unclear about whether resetting a setting resets it to factory default or reverts the user's current changes, which might be exacerbated by some bugs.

Updated tooltip to "Reset setting to default"

> * The interface is no longer as keyboard-friendly.
>   * Keyboard-power-users like myself do fit in a weird space between direct minetest.conf editors and mouse-only operators, but I still edit my settings almost entirely in-game. I know keyboard controls is an afterthought for formspec, I don't think formspecs can even have a focused graphic to distinguish them while tabbing around, but 
>   * The previous interface could basically be entirely navigated by keyboard, with expanding sections and so on, and by tabbing from the search box into the expanding sections. Hitting enter on the search bar would move between search results.
>     * I think the search results should remove the tabbing at left and just put headings in line with the right hand pane. That would be quite keyboard-navigable.
>   *  you have to hold the tab key for a good couple of seconds to tab from the sections at left to the current menu at right.

Yeah, formspecs make it really hard to handle this unfortunately, not sure how to deal with it

> * It would be good if the tooltip text could be activated on a formspec element by hitting space/enter while it's focused, or on the same timer as mouseover.

Tooltips should maybe be shown when focused as well, but not something I can do in this PR. Maybe pressing enter on the info icon can show the info dialog

>   * The info icon has a lesser tab index than the corresponding reset icon for any setting. Thus, instead of tabbing Item-Reset-Info as it appears visually, the tabbing goes Item-Info-Reset.

Fixed

> * I know the existing settings menu doesn't do this, but it would be much better to sort the Mod entries by their displayed name. This would also probably require a different sort conditional on whether show_technical_names is enabled (when that gets re-implemented.

Done based on technical name for now. TODO for titles

>   * Also related to show_technical_names: What about putting the technical name in the tooltip, at least in the interim?

Implemented `show_technical_names`

> * When scrolling with the mousewheel over the left side, if the mouse hovers over a dropdown, it will start scrolling options in that dropdown instead of continuing to scroll. This is exacerbated by the above mentioned bug with the reset buttons on dropdowns not appearing. This is probably not a regression, but this is the single most common interface in minetest where most people will have a scrollbar and dropdowns present.

See #12481 

> **Interface Bugs**
> **Dropdown menus (enum) don't allow returning to defaults correctly:**
> 
>     1. Clear any line that has `language =` in your minetest.conf
>     2. Set language to anything other than "" in-game (I chose `de`) - the reset button will not appear next to the language dropdown as it should.
>     3. Quit the game so the setting is saved. Re-open Minetest. The reset button appears now for the first time.
>     4. Change the language setting to several others, including `""`. The reset button never disappears, even for the `""` setting. Change it back to `""` and close Minetest. You now have an entry `language = ` (yes, that has a trailing space) in your minetest.conf
>     5. Restart Minetest. Now the reset button for language will never appear, which means it is acting the same as in step 2.
>     6. Change the language setting to anything other than `""` again and restart again. Now press the reset button. Now switch it to various languages. No matter what you do with the language setting, the reset button will never appear again.

This is a known issue - I'm currently not triggering a rerender when the dropdown changes due to an issue with reliably getting dropdown actions. So the reset button won't appear until something else causes a render, such as editing another setting

I previously added a TODO to track this, will add to PR description todo list as well

> **Int and float**
> 
>   1. Clear any entry in minetest.conf with `undersampling = `
>   2. Go to the graphics tab. Undersampling will have a value of 0 and a reset button, despite its default and range being 1 1 8 in settingtypes.txt

Undersampling's default is 0 in the .cpp, not sure which default is correct. TODO.

>   3. Type -5 in the field and try to set it. It is clamped to 1. The reset button disappears because this is the default.
>   4. Type -5 in the field again and set it. Close Minetest. The set button won't clamp the value, but the actual value that gets written to minetest.conf will be clamped properly.
>   * Setting an integer setting (I used undersampling) to outside its range is not error-checked except in that when you leave the settings tab it gets clamped to the nearest in-range value.

This is a bug with formspecs in general, it doesn't detect the change so doesn't reset. Added a workaround.

> * Typing anything other than a number in an int/float field and using the set button  will result in a big ugly error message.
> * Entering `NaN` or `-Inf` in an int/float field will convert it to 0 automatically, and clamping to minimum will work from there as well.
> * Entering `Inf` in an int/float field will convert it so it's clamped at the highest possible value for that setting.

Fixed

> **v3f** (I used mapgen fractal scale)
> * Any permutation of capitalisation of `NaN` (`nan`, `NAN`) will get normalised to `nan` and accepted as valid.
> * Any permutation of capitalisation of `Inf` (`inf`, `INF`) will get normalised to `inf` and accepted as valid.
> * Good stuff? Negative zero is normalised to 0. Exponential notation works (e.g. 1e2 = 100).
> * Excessively large positive values such as 1e4096 become inf; excessively large negatives become -inf.

Fixed

> * No feedback if input was invalid. It just ignores everything, doesn't show a reset button, and forgets the invalid value ever existed if you close and reopen Minetest.

Can't reproduce, maybe it was fixed by the work around for formspecs not detecting changes

> **string**
> * The crosshair colour defaults to black if the value is invalid as an RGB triplet, but the default in settingtypes.txt is white.
> * Crosshair colour will dissappear if you attempt something like a formspec injection: `tooltip[0,0;8,8;pwned]`. Thankfully the reset button can fix this easily. I'm not sure if it's possible to do a proper injection since I couldn't get my tooltip to appear.

There is no color setting type currently, so it's just reading it as a string. This is unrelated to this PR, idk whether I should introduce new setting types here

> **bool**
> There is some kind of issue with `bool` with default `false`, but only for mods? Sorry I couldn't reproduce this for any builtin settings such as mip mapping, despite the lines in the respect settingtypes files looking fine...
> 
> 1. Install [foodblocks](https://github.com/Montandalar/foodblocks)
> 2. Remove any `foodblocks_*` settings from `minetest.conf`.
> 3. Start Minetest. Type `food` into the settings search and hit enter.
> 4. Despite being unchecked and the default false, the reset button appears.
> 5. Use the reset button. Toggle the checkbox as many times as you want.
> 7. Once some value is set in `minetest.conf`, the reset button appears as expected.
> 8. Close and reopen Minetest. Checkbox and reset button still work as expected.

Fixed

> **flags**
> The same bug that effects `bool` seems to effect mapgen v6 flags, mapgen v7 flags, carpathian flags, flat flags and valleys flags: if you don't have the default values recorded in minetest.conf, it will offer a reset despite being at the default. Once the reset is pushed it adds the default in and only offers a reset when sensible.

Fixed, although in a brittle way. We need a better API at the engine level for flags

> **filepath**
> I have this version installed through dpkg made by `cpack -G DEB`; it is not a `RUN_IN_PLACE` version.
> 
> Font paths:
> 1. Clear any `font_path_*` settings from `minetest.conf`
> 2. Open Minetest and go to the Advanced settings tab. The regular, bold, italic, bold&italic font paths all have all been set to `/usr/local/share/minetest/fonts/Arimo-*.ttf` and not the suggested resettable path of `fonts/Arimo-Regular.ttf` which is specified in settingtypes.txt. This would not be the correct relative path to the fonts from where my binary is located in `/usr/local/bin/minetest`; and surely it would be `../fonts/` instead for RUN_IN_PLACE builds of Minetest?

This is an engine issue, the settings UI just makes it more obvious

> * Selecting a file path through the file browser is NOT working (I tested 5.5.1 and it works fine)
>   1. Open Minetest and go to the Advanced settings tab
>   2. Click the Browse button next to Regular font path
>   3. Type/paste `/usr/share/fonts/truetype/liberation` (or some other path if you don't have those fonts or are on a different OS) and press enter.
>   4. Click LiberationSerif-Regular.ttf
>   5. Click OK
>   6. Notice that the font path in the field has not actually changed

TODO. Looks like I forgot to implement this

> **noise**
> Accepts Inf, -inf, NaN as valid values. Sets octaves and seed to 0 if invalid. I set most values in the heat noise (mg_biome_np_heat) to NaN and it results no soil at all, probably due to NaNs propagating.
> Could at least defaults, eased, and absvalue, if not all parameters, get tooltips please?

This is unchanged from the current main menu. TODO.

Wuzzy2 commented:
I have tested this, and I am leaving some quick comments and problems I found here:
- Overall, I think this is a big improvement over the complicated tree.
- I don't like the Advanced > Advanced section, it is WAY too large. My biggest problem is the lack of a clear distinction between clientside and serverside settings. Probably at least split this section into two

Some bugs and problems:
- Some numbers are considered "different" from the default value even if it is equal. E.g. if the value entered is "1", it is considered different from the default value "1.0" or vice-versa and the "reset from default" arrow appears.
- The "Edit" button for the mapgen noises should be a bit wider (in German, the translation barely fits)
- You can click the info icon and it behaves like a button. This feels wrong. Better solution IMHO: Turn this into an `image[]` and add a `tooltip[]` in the area of that image. That way, the tooltip will also instantly appear (no delay). I've used this trick in Repixture and it works
- The caption of the text edit fields is sometimes too short. Either in a translation or with the technical name. Examples:  `map_compression_level_net`, `full_block_send_enable_min_time_???`, `max_simultaneaous_block_sends_p???` (the 3 question marks mean I don't know the full name :D )
- Editing the font size is very dangerous and can completely break the GUI. This needs some special care: Set the font size to large or too small, and the interface becomes almost completely unusable and the user cannot trivially back out. There are already caps at 6pt and 72pt, but maybe the limits should be even narrower, especially the upper limit. At 72pt, clicking the widgets often does not work. Or maybe do something like many OSes do when you change the screen resolution: Change the resolution, but pop up a window to force the user to confirm the result within 10 seconds (no response = change the resolution back).
- The info icon is missing for noise settings
- The settings `update_last_checked` and `update_last_known` should not be user-visible at all IMHO, these seem to be for internal use by Minetest only. Instead, add a separate checkbox for disabling update-checking.


Broader stuff for later PRs to keep in mind:
- Overall, the settings as a whole are still quite daunting because there's just so much. I still think the smaller simpler settings screen with the most important settings (+ an "advanced" button for the full list) is the best solution

Stuff that is probably for later PRs:
- A few advanced settings are quite dangerous, I suggest to move them to a separate sub-section under something like "Experimental" or "Danger Zone". I think all settings, including advanced settings should be (mostly) safe to edit unless they are specifically moved into the Danger Zone. Example: `ignore_world_load_errors`, `chunksize`
- The drop-down values should probably be translated. In some cases, this PR actually destroys a translation that already exists, like with `fancy_leaves`
- Mapgen flags should probably be translated as well.
- We shouldn't call mapgen flags "flags" in front of the user. This is too technical. Maybe let's use "modifiers" or something like that
- Each mapgen flag should also have its own info icon instead of a help for all flags together
- In general, I think the GUI currently works very poorly with non-default font-sizes unless you only do a small change. Even at 32pt, most of the text just overflows.
- Missing help for the specific mapgen flags (e.g. `caves`, `dungeons`, etc.)
- I think it would be better to remove the mapgen settings from the big settings menu since they only affect the creation of new worlds (or are there exceptions?). It is more natural and intuitive to move them to the "new world" dialog (perhaps with an "advanced" button or something) because this is the only place where they actually have an effect.
- A few settings like mouse sensitivity and audio volume should be sliders

Problem I have with specific settings (probably PR-unrelated):
- The "3D mode" is confusingly named. Minetest is always 3D! A better term would be "stereoscopy"
- The language selector is confusing. It's not obvious what the empty entry means. Should be rendered as "System default" IMHO
- `invert_mouse` claims it inverts the vertical mouse movement but this is not completely true. It inverts the vertical mouse-movement of the *camera*. In the main menu, it's normal. Important difference!

Probably unrelated problems (that my testing revealed):

- If you click the arrow up or down of the scrollbar or a dropdown, the arrow becomes invisible.
- The tooltip only do a line-break when absolutely neccessary, i.e. when the full width of the screen was used or there is a newline. This can result in very long tooltips if the setting description does not have a manual line-break. Tooltips should line-break automatically earlier; a tooltip should never be as wide as the screen under normal circumstances (example: `debug_log_size_max` in German)

Wuzzy2 commented:
## More bugs!

* The technical names of settings of all Shader settings in Most Used are never shown, even if you enabled technical names
* An error message appears if you edit a text field, no matter if it was a valid or invalid value. Example: `selectionbox_color`

```
Runtime error from mod '' in callback handleMainMenuButtons(): ...test/git/bin/../builtin/mainmenu/settings/components.lua:90: attempt to call upvalue 'converter' (a nil value)
stack traceback:
	...test/git/bin/../builtin/mainmenu/settings/components.lua:90: in function 'on_submit'
	...st/git/bin/../builtin/mainmenu/settings/dlg_settings.lua:487: in function 'handle_buttons'
	/home/wuzzy/src/minetest/git/bin/../builtin/fstk/ui.lua:142: in function 'handle_buttons'
	/home/wuzzy/src/minetest/git/bin/../builtin/fstk/ui.lua:186: in function </home/wuzzy/src/minetest/git/bin/../builtin/fstk/ui.lua:172>
```

* If a setting has no description/tooltip, there is still an info icon next to it (but it is useless). Example: `font_bold`
* Note sure if a bug, but the buttons in the left scroll widget (containing the section names) lack a margin to the right, although they have one to the left

## And now a few general comments about the whole PR:

I am very glad this exists. A rework of the tree structure was desperately needed. The current tree is not great, my biggest problem is that it just has way too many levels and branches. Your approach essentially reduces the "depth" of that tree, which was very important.
The overall structure is good and is far superior to the tree.

The graphics of all this are still very ugly (Minetest default style) but this is obviously PR-unrelated; the layout itself is obviously more important. I still think the main menu needs a more appealing design that look like an actual game thing, not like a tech-demo.

I like the setting descriptions are available as a tooltip this time. However, there is a small issue when the description contains a weblink. It is really inconvenient to remember this, you can't click the weblink nor copy it into the clipboard, you must awkwardly type it by hand.
I also like the easy way to reset any setting to the default and you can see which settings were changed in the first place.
I like that you went with "pixel" icons, overall, this is the style that the whole Minetest main menu should go in the long run (i.e. including all widgets).

What I think could be a bit better is font size. The font size of the sub sections should be larger to emphasize it more. I know the hypertext element supports this, but I didn't check if this is used here. I'm not sure if we can enlarge the font size outside of the hypertext element. :-(

I think it is OK to keep the "technical name" checkbox, although the box around it should be larger (more room for translations). A tooltip for this one is missing because the point of this is not clear.

## About the Most Used section

First of all, there is a glaring omission: Volume. :-)

Also, I think a few things can be safely removed from there since they don't really make sense to toggle in most circumstances:

* `enable_particles` (rationale: I highly doubt many players have a problem with the digging particles)
* `enable_3d_clouds` (rationale: I think flat clouds are not very popular)
* `opaque_water` (rationale: This makes playing at oceans worse, you now must enter the water to see what's inside; IMO turning this on only makes sense for performance reasons)

I suggest to add `show_entity_selectionbox` to the Most Used. It fits well together with `node_highlighting` IMO and opinions vary greatly about whether entities shall have visible selectionboxes.

I suggest to completely hide the other shader settings when the Shader checkbox is turned off. The issue with the grayed-out text is that it weirdly shifts and the dropdown disappears. I know formspecs don't currently support disabled widgets so the closest alternative is to just completely make them disappear. (This should not affect the scroll bar, however, to avoid annoying "jumping")

## Complaints about specific settings (again)

* `show_debug` ended up under "Advanced > Advanced > Network" somehow (should be moved to the developer section instead)
* Move `enable_mapgen_debug_info` from Advanced > Advanced to the developer section
* Move `enable_bloom_debug` to the developer section
* I don't think `arm_inertia` is an accessibility setting, this is purely just eye-candy and a very subtle effect so it should not be in Accessibility
* Add a Camera sub-sub section in Controls > General to keep the camera settings separate from the movement settings
* In Graphics and Sound > Graphics, the settings `view_bobbing_amount` and `fall_bobbing_amount` are about the camera so they should be moved to Camera. And `arm_inertia`, I would move that to the graphical effects (below `enable_particles`)
* Why is `enable_mesh_cache` not enabled by default? Sounds like this could be a nice performance boost for no real drawback
* Why is `connected_glass` off by default?
* In Advanced > Advanced > Server/Environment Performance, a lot of settings are not really all about performance. For example, `abm_interval` can make all ABMs execute more often, which directly affects gameplay (e.g. trees will grow faster). Or `liquid_update` will make liquids flow faster.
* Hide

## List of settings I would move to the Danger Zone (explained in my previous comment)

* `ignore_world_load_errors`
* `chunksize`
* `world_aligned_mode` (experimental)
* `autoscale_mode` (experimental)
* `client_modding` (experimental)
* And possibly all other settings that are officially declared as experimental

rubenwardy commented:
This whole PR is a huge pain because formspecs are so limited. It's very hard to make a good user experience using formspecs, compared to something more capable like HTML/CSS. There are also conflicting goals - Minetest is way too configurable and has way too many settings

Any changes to where settings are placed or what they're named should be done in a separate PR to avoid merge conflicts, as the settingtypes file changes quite a lot. I had noticed that a lot of the settings don't have good names, but didn't want to complicate this PR further. This also includes settings like update_last_checked and update_last_known, these didn't exist when I started this PR - I never added them, they should be in the hidden redundant section at the bottom. Same for the debug (F5). Maybe you'd like to make such a PR?

> Overall, the settings as a whole are still quite daunting because there's just so much. I still think the smaller simpler settings screen with the most important settings (+ an "advanced" button for the full list) is the best solution

The current settings tab is pretty awful, the selection of settings is weird (no language setting?) and it ruins discoverability. This new design should be the default that users use, but I'm not against separating out advanced settings. This could be done using an advanced section, a checkbox, or keeping the advanced dialog in as a separate thing.

> The drop-down values should probably be translated. In some cases, this PR actually destroys a translation that already exists, like with fancy_leaves

This was hard coded in the current settings tab, there's no way to specify translatable options in settingtypes.txt. As a temporary fix, I could hard code the option labels for certain settings to allow them to be translated

If you were to hardcode options, you could use this to fix the `language = ""` issue as well, and get it to show "System default". It would also be nice if the name of the language could be shown rather than the lang code.

> I think it would be better to remove the mapgen settings from the big settings menu since they only affect the creation of new worlds (or are there exceptions?). It is more natural and intuitive to move them to the "new world" dialog (perhaps with an "advanced" button or something) because this is the only place where they actually have an effect.

I agree with this, users should use the settings in the Create World dialog. There are a lot of settings that don't appear in Create World, maybe these should be moved to advanced sections

> A few settings like mouse sensitivity and audio volume should be sliders

For a future PR, there's no slider setting or formspec element (you could use a scrollbar, but not sure that would be understandable)

> If you click the arrow up or down of the scrollbar or a dropdown, the arrow becomes invisible.

I can't reproduce this

> Not sure if a bug, but the buttons in the left scroll widget (containing the section names) lack a margin to the right, although they have one to the left

This is part of the design given by Zughy

> The graphics of all this are still very ugly (Minetest default style) but this is obviously PR-unrelated; the layout itself is obviously more important. I still think the main menu needs a more appealing design that look like an actual game thing, not like a tech-demo.

I could start using Zughy's theme in this PR, but wasn't sure whether it was the best idea to do it before the rest of the menu

The lack of scrollbar styling and the font also ruins it a bit

![image](https://user-images.githubusercontent.com/2122943/197393861-c10434eb-2129-4688-a39c-b6a4766c5732.png)

>  However, there is a small issue when the description contains a weblink. It is really inconvenient to remember this, you can't click the weblink nor copy it into the clipboard, you must awkwardly type it by hand.

There's only one non-advanced setting that has a weblink - contentdb_flag_blacklist. I don't think that is enough to justify adding special handling of this.

> First of all, there is a glaring omission: Volume. :-)

This is redundant with   Pause > Volume

> Also, I think a few things can be safely removed from there

I mostly just tried to copy the existing settings tab

> I don't think arm_inertia is an accessibility setting, this is purely just eye-candy and a very subtle effect so it should not be in Accessibility

Unnecessary movement can cause motion sickness, so being able to turn of such animations is an accessibility thing

Wuzzy2 commented:
Obviously it makes sense to not address everything in this PR, that wasn't my intention anyway, just a couple of stuff to keep in mind for later.

SmallJoker commented:
Moved to 5.8.0 to keep a better overview for the 5.7.0 priorities.
Decision based on https://irc.minetest.net/minetest-dev/2023-01-08#i_6045993

rubenwardy commented:
This is ready for review again

rubenwardy commented:
Most Used should probably be completely removed - I've only kept it there because it has some components to change settings that aren't available elsewhere yet

rubenwardy commented:
It would be good to get this in as soon as possible so it can be iterated on during this release

Desour commented:
Tried out again. My experience, and suggestions for improvements:
* The overall layout is nice. It invites much more to explore all the settings.
* It's not super easy to spot a section in the left list. It would be good to have more pregnant titles (i.e. color them yellow, like in the credits). And icons would also be nice at some point.
* Many setting names are cut-off when used as field title (both, technical names and translated description-names). This is a really severe issue imho.
  One possible fix for this would be to just make the fields much wider.
  Or you could also add the (technical or descriptive, whatever is currently selected) name as a tooltip to each `field[]` element.
  You could also add the setting name as title to the tooltips of the info / (i) elements.
* I'd like to get old setting change menu if I click on an (i). (For clarity: I'm referring to the menu that you get after you click on "change" in the current old advanced settings tree.)
  Reasons:
    * Reading a long setting description in a tooltip is kinda weird, might not work in small windows, and is hard because the tooltip is far on the right of the screen.
    * Idk how the tooltips work on touch devices. But if you have to hold a finger down, it's probably quite annoying.
    * If some settings will get sliders, people may still want to manually enter values. The old setting change menu could provide this.
    * The old setting change menu can also show the full setting name, in case it got cut off earlier.

Desour commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,510 @@
+
+	for i, comp in ipairs(dialogdata.components) do
+		if comp.on_submit and comp:on_submit(fields, this) then
+			print("Event handled by component ", dump(comp))
```
Note: Dubug print is still there.

rubenwardy commented:
> It's not super easy to spot a section in the left list. It would be good to have more pregnant titles (i.e. color them yellow, like in the credits). And icons would also be nice at some point.

Done

![image](https://user-images.githubusercontent.com/2122943/230974380-011945fb-04ea-4389-849a-4a4fedaa66c2.png)

> Many setting names are cut-off when used as field title (both, technical names and translated description-names). This is a really severe issue imho.

Any examples? Are you using a non-default font size / gui scaling? It all fits for me without clipping at default screen size

> I'd like to get old setting change menu if I click on an (i). (For clarity: I'm referring to the menu that you get after you click on "change" in the current old advanced settings tree.)

Obviously the (i) button should be made to work on Android - adding a dialog for this is fine by me. But I don't think there's a need for a the old setting change dialog, for things like slider values I'd rather have that inline as a slider component

Desour commented:
>Any examples? Are you using a non-default font size / gui scaling? It all fits for me without clipping at default screen size

I'm using default scaling settings. (It gets very slightly less bad when I maximize my window though.)
Examples:
![2023-04-10-212335_1920x1080_scrot](https://user-images.githubusercontent.com/7613443/230981608-76f2dc6d-fbe4-4ab2-8349-7a5885c941fe.png)

rubenwardy commented:
Fixed by making the fields use the full available width

binarydigitz01 commented:
I like this PR. Just wanted that the mouse sensivity should be able to be adjusted via a slider. That is more user friendly and is easier to adjust.

rubenwardy commented:
> I like this PR. Just wanted that the mouse sensivity should be able to be adjusted via a slider. That is more user friendly and is easier to adjust.

Thanks! Added that to the future improvements list, I was planning on doing that in the future but hadn't noted it down

v-rob commented on builtin/mainmenu/settings/components.lua:
```diff
@@ -0,0 +1,378 @@
+		end,
+
+		on_submit = function(self, fields)
+			if fields["set_" .. setting.name] then
```
```suggestion
			if fields["set_" .. setting.name] or
					fields.key_enter_field == setting.name .. "_x" or
					fields.key_enter_field == setting.name .. "_y" or
					fields.key_enter_field == setting.name .. "_z" then
```
Again, if we want to allow Enter key.

v-rob commented on builtin/mainmenu/settings/components.lua:
```diff
@@ -0,0 +1,378 @@
+				end
+				return true
+			end
+			if fields["set_" .. setting.name] then
```
```suggestion
			if fields["set_" .. setting.name] or fields.key_enter_field == setting.name then
```
Enter again

v-rob commented on builtin/mainmenu/settings/components.lua:
```diff
@@ -0,0 +1,378 @@
+			if fields["pick_" .. setting.name] then
+				local is_file = setting.type ~= "path"
+				core.show_path_select_dialog(dialog_name,
+					is_file and fgettext_ne("Select file") or fgettext_ne("Select directory"), is_file)
```
```suggestion
					is_file and fgettext("Select file") or fgettext("Select directory"), is_file)
```

rubenwardy replied:
`core.show_path_select_dialog` is not a formspec function so should not use `formspec_escape`

v-rob replied:
Ah, of course

v-rob commented on builtin/mainmenu/settings/components.lua:
```diff
@@ -0,0 +1,378 @@
+
+					local is_checked = checkboxes[possible]
+					fs[#fs + 1] = ("checkbox[%f,%f;%s;%s;%s]"):format(
+						x, y + 0.25, setting.name .. "_" .. possible,
```
If you replace the `local y = 0.3` above with `local y = 0.55`, then this `y + 0.25` is unnecessary.

v-rob commented on builtin/mainmenu/settings/components.lua:
```diff
@@ -0,0 +1,378 @@
+					end
+				end
+
+				core.settings:set(setting.name,  table.concat(values, ","))
```
```suggestion
				core.settings:set(setting.name, table.concat(values, ","))
```

v-rob reviewed (commented):
Halfway through reviewing the PR, and have done a little bit of preliminary testing.  I really love how much nicer this settings GUI is!  Thanks for your hard work.

Two things not in the review comments (I didn't read many of the previous comments, so apologies if I repeat anything anyone else said):

* Input fields don't allow pressing enter.  I made suggestions to fix this.
* Colors in the format `(r, g, b)` allow wacky values like `(0, asd, NaN)`, which is out of scope for the PR to fix since there is no setting type for colors AFAICT.  Hex/named color notation à la ColorString would be far nicer anyway.

v-rob commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+
+
+local function get_shadow_mapping_id()
+	local shadow_setting = core.settings:get("shadow_levels")
+	for i = 1, #shadow_levels_options[2] do
+		if shadow_setting == shadow_levels_options[2][i] then
+			return i
+		end
+	end
+	return 1
```
```suggestion
	local level = tonumber(core.settings:get("shadow_levels"))
	if level and level >= 0 and level < #shadow_levels_labels then
		return level + 1
	end
	return 1
```
I'm pretty sure this is equivalent--I don't see why we need an array of stringified numbers.

v-rob reviewed (commented):
Another batch of reviewing.  One more file to go after this.

rubenwardy commented:
Please note that `shader_component.lua` is just code from the existing `tab_settings.lua` updated to the new GUI

v-rob commented:
Bah!  More code I didn't have to review!  However, positions still must be updated.

v-rob commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,514 @@
+		end
+
+		if comp.info_text and comp.info_text ~= "" then
+			used_h = math.max(used_h or 0, 0.5)
```
This implies that `used_h` may be nil, but `used_h` is used before this line.  Can it actually be nil?  If so, this should be moved up.

v-rob commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,514 @@
+
+		fs[#fs + 1] = "container_end[]"
+
+		if used_h and used_h > 0 then
```
Same here about `used_h`.

v-rob commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,514 @@
+			used_h = math.max(used_h or 0, 0.5)
+
+			local info_x = right_pane_width - 0.75
+			local info_y = used_h / 2 - 0.25
```
This `info_y` calculation is duplicated for this and the reset button.  Move the variable up?

v-rob reviewed (commented):
Finished reviewing.  If the comments are addressed and my testing turns up no other bugs, this PR has my approval.

sfan5 commented:
I absolutely can't find the time to review this. If @v-rob is finished with his and fine with the PR please merge it soon.

v-rob commented:
Yeah, as long as the review comments are addressed, it's got my approval.  There's a bunch of them, but they're all pretty minor issues and won't take much modification to fix.  Only a few of them are actually bugs, and the others are just small code quality requests.

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+		if shaders_enabled then
+			fs = fs ..
+				"container[0,0.75]" ..
+				"checkbox[0,0;cb_tonemapping;" .. fgettext("Tone Mapping") .. ";"
```
```suggestion
				"checkbox[0,0;cb_tonemapping;" .. fgettext("Tone mapping") .. ";"
```

Iniquitatis replied:
```suggestion
				"checkbox[0,0;cb_tonemapping;" .. fgettext("Tone mapping") .. ";"
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+				"container[0,0.75]" ..
+				"checkbox[0,0;cb_tonemapping;" .. fgettext("Tone Mapping") .. ";"
+						.. tostring(core.settings:get_bool("tone_mapping")) .. "]" ..
+				"checkbox[0,0.5;cb_waving_water;" .. fgettext("Waving Liquids") .. ";"
```
```suggestion
				"checkbox[0,0.5;cb_waving_water;" .. fgettext("Waving liquids") .. ";"
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+						.. tostring(core.settings:get_bool("tone_mapping")) .. "]" ..
+				"checkbox[0,0.5;cb_waving_water;" .. fgettext("Waving Liquids") .. ";"
+						.. tostring(core.settings:get_bool("enable_waving_water")) .. "]" ..
+				"checkbox[0,1;cb_waving_leaves;" .. fgettext("Waving Leaves") .. ";"
```
```suggestion
				"checkbox[0,1;cb_waving_leaves;" .. fgettext("Waving leaves") .. ";"
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+
+			if video_driver == "opengl" then
+				fs = fs ..
+					"label[0,2.2;" .. fgettext("Dynamic shadows:") .. "]" ..
```
```suggestion
					"label[0,2.2;" .. fgettext("Dynamic shadows") .. "]" ..
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+			fs = fs ..
+				"container[0.35,0.75]" ..
+				"label[0,0;" .. core.colorize("#888888",
+						fgettext("Tone Mapping")) .. "]" ..
```
```suggestion
						fgettext("Tone mapping")) .. "]" ..
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+				"label[0,0;" .. core.colorize("#888888",
+						fgettext("Tone Mapping")) .. "]" ..
+				"label[0,0.5;" .. core.colorize("#888888",
+						fgettext("Waving Liquids")) .. "]" ..
```
```suggestion
						fgettext("Waving liquids")) .. "]" ..
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+				"label[0,0.5;" .. core.colorize("#888888",
+						fgettext("Waving Liquids")) .. "]" ..
+				"label[0,1;" .. core.colorize("#888888",
+						fgettext("Waving Leaves")) .. "]" ..
```
```suggestion
						fgettext("Waving leaves")) .. "]" ..
```

Iniquitatis commented on builtin/mainmenu/settings/shader_component.lua:
```diff
@@ -0,0 +1,160 @@
+				"label[0,1;" .. core.colorize("#888888",
+						fgettext("Waving Leaves")) .. "]" ..
+				"label[0,1.5;" .. core.colorize("#888888",
+						fgettext("Waving Plants")) .. "]"..
```
```suggestion
						fgettext("Waving plants")) .. "]"..
```

Iniquitatis reviewed (commented):
(Nitpicks about inconsistencies with almost every other setting label.)

Iniquitatis commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,514 @@
+
+local full_settings = settingtypes.parse_config_file(false, true)
+local info_icon_path = defaulttexturedir .. "settings_info.png"
+local reset_icon_path = defaulttexturedir .. "settings_reset.png"
```
```suggestion
local reset_icon_path = core.formspec_escape(defaulttexturedir .. "settings_reset.png")
```
And the same with `info_icon_path`.
Otherwise the icons break if Minetest is located somewhere like `C:\Users\Blah\Downloads\msvc-x64-portable\minetest-5.8.0-dev-win64\bin`.
(Not sure if this is the case because of those Windows backslashes, or something else.)

**Before:**
![image](https://user-images.githubusercontent.com/5585984/234787497-b71083b2-9740-4bc2-ba02-cc811c9ab974.png)

**After:**
![image](https://user-images.githubusercontent.com/5585984/234789064-40273690-faaa-430b-a6b5-903ff39311f3.png)

rubenwardy commented on builtin/mainmenu/settings/dlg_settings.lua:
```diff
@@ -0,0 +1,514 @@
+	local retval = {}
+	local j = 1
+	for i, content in ipairs(page.content) do
+		if content == false then
+			-- skip
+			j = j - 1
```
This is used to hude `autosave_screensize` on Android, and other stuff like that

```lua
PLATFORM ~= "Android" and "autosave_screensize" or false,
```

rubenwardy commented:
@v-rob updated according to review

lhofhansl commented:
Seeing this now upon start: @rubenwardy 

```
2023-05-01 10:30:57: ERROR[Main]: GUIEngine: execution of menu script failed: Failed to load and run script from /opt/minetest/share/minetest/builtin/init.lua:
2023-05-01 10:30:57: ERROR[Main]: ...hare/minetest/builtin/mainmenu/settings/dlg_settings.lua:42: Page content:_mods_lightning already registered
2023-05-01 10:30:57: ERROR[Main]: stack traceback:
2023-05-01 10:30:57: ERROR[Main]:       [C]: in function 'assert'
2023-05-01 10:30:57: ERROR[Main]:       ...hare/minetest/builtin/mainmenu/settings/dlg_settings.lua:42: in function 'add_page'
2023-05-01 10:30:57: ERROR[Main]:       ...hare/minetest/builtin/mainmenu/settings/dlg_settings.lua:147: in function 'load_settingtypes'
2023-05-01 10:30:57: ERROR[Main]:       ...hare/minetest/builtin/mainmenu/settings/dlg_settings.lua:161: in main chunk
2023-05-01 10:30:57: ERROR[Main]:       [C]: in function 'dofile'
2023-05-01 10:30:57: ERROR[Main]:       ...netest/share/minetest/builtin/mainmenu/settings/init.lua:22: in main chunk
2023-05-01 10:30:57: ERROR[Main]:       [C]: in function 'dofile'
2023-05-01 10:30:57: ERROR[Main]:       /opt/minetest/share/minetest/builtin/mainmenu/init.lua:43: in main chunk
2023-05-01 10:30:57: ERROR[Main]:       [C]: in function 'dofile'
2023-05-01 10:30:57: ERROR[Main]:       /opt/minetest/share/minetest/builtin/init.lua:63: in main chunk
2023-05-01 10:30:57: ERROR[Main]: No future without main menu!
```

runsy commented:
I've been testing it. And my first impression: HORRENDOUS.

- It's horrendous: Please, one click to access the settting. Now you have to double click, tab and button to access, this is counterintuitive.
My idea: I miss the options of before, few but the most used, shadows, quality of the leaves, etc. Now I don't know where the shadows are activated, I had to spend 5 min to find them. So as before, those advanced options through, and the advanced options button. And keep the previous page menu. That is, "the most used" in the "Setings" tab, together with the "advanced button", as before.
- Then, all the settings appear: The inexperienced user will get lost. And not to mention the children, what? Look at the KODI options menu.
I propose: Several settings options, maybe thru buttons: basic, advanced and expert. What do I and 99% care about the "water_wave_speed" option? 
-Aesthetic aspects: Icons (pixel art) in the left panel. Too much text and variables as is that it is difficult to decipher.

runsy commented:
Another thing:

If you put a side panel, the window should not be square but wider than high. So aesthetically and ergonomically add 2 slots width to the settings window.

rubenwardy commented:
The current theming is temporary, as is the "open settings" button. See [this image](https://user-images.githubusercontent.com/2122943/197393861-c10434eb-2129-4688-a39c-b6a4766c5732.png) for what it might look like in the future

> Sorry Karl Lagerfeld would be an apprentice compared to me in matters of good taste and aesthetics. :-D

This you?

![logo](https://github.com/minetest/minetest/assets/2122943/b05f4be3-40bc-486a-8c7e-fe26a6a30880)

runsy commented:
> The current theming is temporary, as is the "open settings" button. See [this image](https://user-images.githubusercontent.com/2122943/197393861-c10434eb-2129-4688-a39c-b6a4766c5732.png) for what it might look like in the future
> 
> Remember that this is just the first iteration, this PR was huge enough already
> 
> > My idea: I miss the options of before, few but the most used, shadows, quality of the leaves, etc. Now I don't know where the shadows are activated, I had to spend 5 min to find them.
> 
> This is still the first page you get to - most used - but we're planning on removing stuff and instead having things like the shadows drop-down in the graphics settings. So to enable shadows you'd just go to graphics > shaders
> 
> > I propose: Several settings options, maybe thru buttons: basic, advanced and expert. What do I and 99% care about the "water_wave_speed" option?
> 
> This is planned: #13476

I see. Sorry, I thought this was already the definitive version. With those changes much better.
